Access control updates (#1499)
semgrep/semgrep-docs: docs/semgrep-cloud-platform/teams.md

---
slug: teams
append_help_link: true
title: Manage access to projects
hide_title: true
description: Use the Teams feature to manage access to resources.
tags:
  - Semgrep Cloud Platform
---

# Manage access to projects

Use the **Teams** feature to manage access to resources, such as findings in your projects.

![The Teams tab within the Settings page](/img/access-teams.png)
**Figure**. The **<i class="fa-solid fa-gear"></i> Settings > Access > Teams** tab displays both top-level teams and subteams.

> The **Teams** feature enables admins to grant or limit access to specific projects in Semgrep Cloud Platform (SCP). **Projects** are repositories or codebases you have added to SCP for scanning. You can quickly assign projects to large groups of members by assigning teams and subteams to members of your organization.

Reviewer: ```suggestion
The **Teams** feature enables admins to grant or limit access to specific projects in Semgrep Cloud Platform (SCP). **Projects** are repositories or codebases you have added to SCP for scanning. You can quickly assign projects to large groups of members by first assigning them to teams and subteams within your organization.
```

IDK if that's the greatest way of saying that, but the original sentence was a bit confusing to me

This feature helps security engineers and developers in large organizations focus on the projects that are relevant to their specific department or team.

> When you limit a member's access to a subset of your projects, their Dashboard and Findings pages all reflect that change. For example, their total finding count is based on the total number of findings of the projects they can access.

Reviewer: ```suggestion
When you limit a member's access to a subset of your projects, their **Dashboard** and **Findings** pages all reflect that change. For example, their total finding count is based on the total number of findings of the projects they can access.
```

## Roles and visibility

A user is any person who has been added to your organization in Semgrep. Users can be administrators (admins), managers, or members.

<dl>
<dt>Admin</dt>
<ul><li>A user who has access to all features, resources, and projects of their Semgrep deployment. Admins can also change the role of members and managers.</li>
<li>By definition, an admin does not need to be assigned to a team to have access to the resources of a project. As such, an admin can't be removed from a team.</li>
<li>An org admin can change the role of any other admin or user, including a fellow admin.</li></ul>
<dt>Manager</dt>
<ul>
<li>A user who can grant access to projects by creating subteams and assigning members to these subteams.</li>
<li>A manager role is restricted to the teams where they have been assigned as a manager. Users can be managers of some projects, but members for others. For more information, see the <a href="#the-manager-role">manager role</a>.</li>
</ul>
<dt>Member</dt>
<ul>
<li>A user who has access to some features, resources, and projects of their Semgrep deployment.</li>
<li>By default, a member starts out with no access to any project.</li>

> <li>To grant members access to a project and its findings, you must add a member to a team, and that team must be assigned projects.</li>

Reviewer: ```suggestion
<li>To grant members access to a project and its findings, you must add the members to a team, and that team must be assigned to the project.</li>
```

<li>Members can scan their local or personal repositories through a personal account.</li>
</ul>
</dl>

![A member's view of the Projects page.](/img/access-member-view.png)
**Figure**. A member's view of the Projects page. It displays projects that are assigned to the team they are a member of, but they cannot edit a project nor can they scan new projects in their organizational account.

> 

Reviewer: This is a completely optional suggestion, but maybe we could add headers to the tables? It would split them up a bit and allow people to link directly to the tables.

| Page      | Member     | Manager    | Admin | Notes |
| ----      | ------     | -------    | ----- | ----- |

> | Dashboard | ⚠️&nbsp;Restricted | ⚠️&nbsp;Restricted | ✅ Yes   | For members, scope is limited based on a member's teams and project access granted to those teams. |

Reviewer: ```suggestion
| Dashboard | ⚠️&nbsp;Restricted | ⚠️&nbsp;Restricted | ✅ Yes   | For members, scope is limited based on a their teams and the project access granted to those teams. |
```

| Delete subteams         | ❌ No     | ✅ Yes     | ✅ Yes   | A manager can delete the subteams they are a manager of, provided that there are no resources, such as projects, assigned to the subteam.      |
| API                     | ❌ No     | ❌ No      | ✅ Yes   |       |

## How team access works

- Members of a top-level team gain access to the projects of its subteams. They are indirect members of a subteam.
- Members of a subteam do not have access to the projects of teams or subteams above it.

In the following diagram, team 1 gains access to subteam 1b's projects, but team 1b does not gain access to projects from team 1.

![Team scopes diagram](/img/access-diagram.png#md-width)

- The members Alexis, Pam, and Raj have access to the following projects:
    - App
    - Microservices
    - Frontend
- The members David, Sebas, and Phaedra have access to the following projects:
    - Frontend

### The manager role

Use the **manager role** to delegate the assignment of projects across many users. Managers can speed up the deployment of Semgrep into your organization by creating subteams to grant members access to projects.

> Given a security engineer who is a manager for **team A** but member for **team B**, with both teams having the same projects:

Reviewer: ```suggestion
Given a security engineer who is a manager of **team A** but a member of **team B**, with both teams having the same projects:
```

- The security engineer has manager **access** to the projects.

> - The security engineer can't create subteams for team B but can create subteams for team A.

Reviewer: ```suggestion
- The security engineer can create subteams for team A but can't create subteams for team B.
```

## Tips for creating teams and subteams

:::info
The Semgrep team recommends that admins assign projects to one team only.
:::

- **Use subteams to grant access to a specific department's repositories**: Create a top-level team for managers or security engineers in your organization who have broad access to a variety of repositories, then create subteams for members to grant them limited access to their specific department's repositories.
- **Use flat teams to grant access to central projects that are used by a broad group of developers**: It is best to create a separate flat team, without any subteams, and grant the users access to foundational or central repositories from that team. For example, projects that all engineers commit to can be named the Engineering Team.

## Configure your teams

### View your teams

You must be an admin or manager to view the **Teams** tab.

1. Sign in to [<i class="fas fa-external-link fa-xs"></i> Semgrep Cloud Platform](https://semgrep.dev/login).
1. Click **[<i class="fa-solid fa-gear"></i> Settings > Access > Teams](https://semgrep.dev/orgs/-/settings/access/teams)**.

### Create a team

> 1. Click **New team**. The **Create New Team** form appears.

Reviewer: (This applies to all of the instructions sections) Does it make sense to include the log in and the Settings > Access > Teams instructions at the top of each list just in case someone lands directly at that section and not at the H2?

Author: We could do something like the [Policies page](http://semgrep.dev/docs/semgrep-code/policies/), in which we say "In the Settings > Access > Teams tab" and provide a link to the tab. I'll add that in.

---
slug: teams
append_help_link: true
title: Manage access to projects
hide_title: true
description: Use the Teams feature to manage access to resources.
tags:
  - Semgrep Cloud Platform
---

# Manage access to projects

Use the **Teams** feature to manage access to resources, such as findings in your projects.

![The Teams tab within the Settings page](/img/access-teams.png)
**Figure**. The **<i class="fa-solid fa-gear"></i> Settings > Access > Teams** tab displays both top-level teams and subteams.

> The **Teams** feature enables admins to grant or limit access to specific projects in Semgrep Cloud Platform (SCP). **Projects** are repositories or codebases you have added to SCP for scanning. You can quickly assign projects to large groups of members by first assigning them to teams and subteams within your organization.

Author: @khorne3 I guess we're reverting https://github.com/semgrep/semgrep-docs/pull/1499#discussion_r1578187816 :P

This feature helps security engineers and developers in large organizations focus on the projects that are relevant to their specific department or team.

> When you limit a member's access to a subset of your projects, their **Dashboard** and **Findings** pages all reflect that change. For example, their total finding count is based on the total number of findings of the projects they can access.

Author: User = everyone including admins
Member = not an admin

This document walks you through the following:

- How to approach team management and project access in Semgrep
- How to create, view, update, and delete teams and subteams
- How to assign or unassign projects to teams

## Roles and visibility

A user is any person who has been added to your organization in Semgrep. Users can be administrators (admins), managers, or members.

<dl>
<dt>Admin</dt>
<ul><li>A user who has access to all features, resources, and projects of their Semgrep deployment. Admins can also change the role of members and managers.</li>

> <li>By definition, an admin does not need to be assigned to a team to have access to the resources of a project. As such, an admin can't be removed from a team.</li>

Author: > An admin can't be removed from the team

This is based on the UI - the admins are all there but are greyed out (can't be removed.) I will make that clearer.

> An admin does not need to be assigned to a team

Also based on the way the UI behaves - the admins are there, but "nobody" explicitly added them.

> <li>An org admin can change the role of any other admin or user, including a fellow admin.</li></ul>

Author: Good catch. Maybe admin, member, or manager.

<dt>Manager</dt>
<ul>
<li>A user who can grant access to projects by creating subteams and assigning members to these subteams.</li>
<li>A manager role is restricted to the teams where they have been assigned as a manager. Users can be managers of some projects, but members for others. For more information, see the <a href="#the-manager-role">manager role</a>.</li>
</ul>

> <dt>Member</dt>

Author: Hah. I think this is also a by-product of Teams being on top of the original user management. The very top of the doc actually makes that pretty clear:

![image](https://github.com/semgrep/semgrep-docs/assets/20766840/da65eb00-4944-459e-a8bb-311d0e3c82b5)

I'll adapt that for this section as well.

> <dt>Member</dt>

Author: Quick render that addresses the last 3 comments:

![image](https://github.com/semgrep/semgrep-docs/assets/20766840/d2b3b6a7-c54d-41c9-b169-204e28379658)

> <dt>Member</dt>

Reviewer 2: LGTM 😄

---
slug: teams
append_help_link: true
title: Teams and user management
hide_title: true
description: Manage user access to projects through Teams.
tags:
  - Semgrep Cloud Platform
---

import MoreHelp from "/src/components/MoreHelp"

# Manage user access to projects

Use the **Access** page to manage access to Semgrep resources, such as scans, projects, and findings. **Projects** are repositories or codebases you have added to SCP for scanning.

## User roles and access

Accounts enable you to manage access to Semgrep resources, such as scans, projects, and findings, with varying levels of collaboration and visibility. Basic access control is managed through the **Members** tab.

![Role-based access control](/img/rbac-overview.png)<br />

Semgrep primarily divides users into two roles:

* `admin`
* `member`

Optionally, you can assign members to a third role, **manager**, for more granular assignment of user access. Managers are a subset of members with some additional capabilities and scopes.

:::info

> * Users are assigned a role based on your [organization's default](#setting-a-default-role). New organizations are created with a default role of `admin`.

Author: I think this is where a render would be useful:

![image](https://github.com/semgrep/semgrep-docs/assets/20766840/fae20513-39ae-46c5-956f-47b2d0d76704)

And here's the section with the choices for defaults:

![image](https://github.com/semgrep/semgrep-docs/assets/20766840/2a076218-3c28-4187-a072-aa7ea22501b0)

> * Users are assigned a role based on your [organization's default](#setting-a-default-role). New organizations are created with a default role of `admin`.

Reviewer 2: ah, indeed. thanks! sorry about that

</dl>

![A member's view of the Projects page.](/img/access-member-view.png)
**Figure**. A member's view of the Projects page. It displays projects that are assigned to the team they are a member of, but they cannot edit a project nor can they scan new projects in their organizational account.

### Page visibility per role

| Page      | Member     | Manager    | Admin | Notes |
| ----      | ------     | -------    | ----- | ----- |
| Dashboard | ⚠️&nbsp;Restricted | ⚠️&nbsp;Restricted | ✅ Yes   | For members, scope is limited based on a their teams and the project access granted to those teams. |
| Projects  | ⚠️&nbsp;Restricted | ⚠️&nbsp;Restricted | ✅ Yes   | Projects assigned to teams are visible to members and managers of those teams. Admins can see all projects. |
| Findings  | ⚠️&nbsp;Restricted | ⚠️&nbsp;Restricted | ✅ Yes   | Members can perform all triage operations on Projects assigned to them. |
| Policies  | ❌ No         | ❌ No         | ✅ Yes   |       |
| Editor    | 👁️&nbsp;Read-only  | 👁️&nbsp;Read-only  | ✅ Yes   | Members can view all rules of an organization, but can't edit or create their own. They can create their own rules in their personal account.      |

> | Settings  | ❌ No         | ⚠️&nbsp;Restricted | ✅ Yes   |       |

Reviewer 3: Probably need to add something here about what managers can do on the settings page. I believe they can see the Account tab and the Access tab. On the Access tab they can only see the Members and Teams tabs. I'm not sure what teams and users they can see. I think they can see only the teams they manage but they can see all users?

Reviewer 3: Just kidding - they can see all teams but only edit theirs. And they can see all users but not all projects

Author: @sabrinabrogren iirc they can't see the **Members** tab but can see the **Teams** tab, is that right?

Author: Sorry, you already answered it:
>  On the Access tab they can only see the Members and Teams tabs.

Given a security engineer who is a manager of **team A** but a member of **team B**, with both teams having the same projects:

- The security engineer has manager **access** to the projects.
- The security engineer can create subteams for team A but can't create subteams for team B.

Additionally, the manager role is able to perform the following:

- Scan new projects through the **Projects** page.
- Edit projects that their team is assigned to.

Managers cannot remove themselves from their team. Only admins can remove managers.

![A manager's view of the Projects page.](/img/access-manager-view.png)
**Figure**. A manager's view of the Projects page. They are able to scan new projects and edit the settings for Projects assigned to Teams they are managers of.

## Enable teams

1. Sign in to [<i class="fas fa-external-link fa-xs"></i> Semgrep Cloud Platform](https://semgrep.dev/login).
1. Click **[<i class="fa-solid fa-gear"></i> Settings > Access > Teams](https://semgrep.dev/orgs/-/settings/access/teams)**.
1. Click **Enable**.

> When you have enabled teams for the first time, a team is automatically created with the name of your deployment. This preserves the settings you previously had using the **Members** feature; All current members retain their existing projects.

Reviewer 2: We recently added a feature that optionally allows the default team to also include newly added members. With this option enabled, enabling Teams will preserve the previous behavior (new members will have access to every project in the default team, which we create with all projects). It might be worth highlighting that here.

Reviewer 2: Hmm I don't think there is a figma, we sort of winged it. There's a screenshot in https://github.com/semgrep/semgrep-app/pull/13923
